src/thor/route_matcher.cc

```diff
@@ -235,73 +237,85 @@ bool expand_from_node(const sif::mode_costing_t& mode_costing,
+  if (options.shape_size() < 2) {
     throw std::runtime_error("Invalid shape - less than 2 points");
   }
 
   // Form distances between shape points and accumulated distance from start to each shape point
   float total_distance = 0.0f;
   std::vector<std::pair<float, float>> distances;
   distances.push_back(std::make_pair(0.0f, 0.0f));
-  for (size_t i = 1; i < shape.size(); i++) {
-    float d = shape[i].lnglat().Distance(shape[i - 1].lnglat());
+  for (size_t i = 1; i < options.shape_size(); i++) {
+    float d = to_ll(options.shape(i).ll()).Distance(to_ll(options.shape(i - 1).ll()));
     total_distance += d;
     distances.push_back(std::make_pair(d, total_distance));
   }
 
   // Keep a record of followed edges and transition from each shape index (for each hierarchy level) -
   // this prevents doubling back and causing an infinite loop (could be due to transitions)
   followed_edges_t followed_edges;
-  followed_edges.resize(shape.size());
+  followed_edges.resize(options.shape_size());
   for (auto& f : followed_edges) {
     f.resize(TileHierarchy::get_max_level());
   }
 
   // Process and validate end edges (can be more than 1). Create a map of
   // the end edges' start nodes and the edge information.
+  // TODO: when we want to do more than one leg we need to create correlated path_edges on the fly
   auto end_nodes = GetEndEdges(reader, options.locations());
 
   // We support either the epoch timestamp that came with the trace point or
   // a local date time which we convert to epoch by finding the first timezone
-  graph_tile_ptr tile = nullptr;
-  const DirectedEdge* directededge = nullptr;
-  const NodeInfo* nodeinfo = nullptr;
-  uint32_t origin_epoch = 0;
-  for (const auto& e : options.locations().begin()->path_edges()) {
-    GraphId edgeid(e.graph_id());
-    if (!edgeid.Is_Valid() || !reader.GetGraphTile(edgeid, tile))
-      continue;
-    directededge = tile->directededge(edgeid);
-    if (reader.GetGraphTile(directededge->endnode(), tile)) {
-      // get the timezone
-      nodeinfo = tile->node(directededge->endnode());
-      const auto* tz = DateTime::get_tz_db().from_index(nodeinfo->timezone());
-      if (!tz)
-        continue;
-      // if its timestamp based need to signal that out to trip leg builder
-      if (!options.shape(0).has_date_time() && options.shape(0).time() != -1.0) {
-        options.mutable_shape(0)->set_date_time(
-            DateTime::seconds_to_date(options.shape(0).time(), tz, false));
-      }
-      // remember where we are starting
-      if (options.shape(0).has_date_time()) {
-        if (options.shape(0).date_time() == "current")
-          options.mutable_shape(0)->set_date_time(DateTime::iso_date_time(tz));
-        origin_epoch = DateTime::seconds_since_epoch(options.shape(0).date_time(), tz);
-      }
-      break;
-    }
-  }
+  valhalla::baldr::DateTime::tz_sys_info_cache_t tz_cache;
+  auto time_info = init_time_info(reader, options, &tz_cache);
```

Comment:
so for the shape walking map matcher (ie the faster less robust one) we also switch to using time_info